python/ql/src/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -1268,11 +1305,38 @@ module UnpackingAssignment {
     ControlFlowNode getAnElement() { result = this.getElement(_) }
   }
 
+  /**
+   * Step 1a
+   * Data flows from `iterable` to `TIterableSequence(sequence)`
+   */
+  predicate iterableUnpackingAssignmentFlowStep(Node nodeFrom, Node nodeTo) {
+    exists(AssignmentTarget target |
+      nodeFrom.asExpr() = target.getValue() and
+      nodeTo = TIterableSequenceNode(target)
+    )
+  }
+
+  /**
+   * Step 1b
+   * Data is read from `iterable` to `TIterableSequence(sequence)`
+   */
+  predicate iterableUnpackingForReadStep(CfgNode nodeFrom, Content c, Node nodeTo) {
+    exists(ForTarget target |
+      nodeFrom.asExpr() = target.getSource() and
+      nodeTo = TIterableSequenceNode(target.(SequenceNode))
```

Comment:
This cast is sort of awkward. Would it be possible to move this restriction to `ForTarget` itself?
Ah, I guess may not, since it would only work with inheritance...

This is the second time things have been a bit awkward since `SequenceNode` is abstract. Perhaps it's time to use a `Range` pattern for this?

Reply:
`ForTarget` is also used in the regular `for`-step, where it is not a `SequenceNode`. The cast here completely captures my intent (I want to create a `TIterableSequenceNode`, but only if `target` is a `SequenceNode`). We could write it explicitly with `instanceof`?

Reply:
> This is the second time things have been a bit awkward since `SequenceNode` is abstract. Perhaps it's time to use a `Range` pattern for this?

I don't think we should use the `::Range` pattern, since we don't want outside users to extend it further. Instead I think we should write it like [`IterableNode`](https://github.com/github/codeql/blob/9f50f67e6d6abde62312e5b444604e953d6df113/python/ql/src/semmle/python/Flow.qll#L780-L784)

Reply:
> `ForTarget` is also used in the regular `for`-step, where it is not a `SequenceNode`. The cast here completely captures my intent (I want to create a `TIterableSequenceNode`, but only if `target` is a `SequenceNode`). We could write it explicitly with `instanceof`?

Yes, I think that would be good :+1:

```suggestion
      target instanceof SequenceNode
      nodeTo = TIterableSequenceNode(target)
```

Reply:
> I don't think we should use the `::Range` pattern, since we don't want outside users to extend it further.

This seems like a non-sequitur to me. There's no rule that says the `Range` pattern _must_ be used for the sole purpose of allowing for extensions by external users. (For instance, we could make the `Range` part private.) The JavaScript library has lots of uses of the `Range` pattern in places where I don't think the ultimate goal is to allow easy extension by _external_ users, but rather to allow a more modular way of writing code for the _internal_ users. See e.g. [`PartialInvokeNode`](https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/dataflow/Nodes.qll#L1282) where, _yes_, you could write this together as a single class, but it would be an incomprehensible mess.

(Also, how do we know that users will never need to extend `SequenceNode`? Maybe they have a _really weird_ way of encoding sequences that we don't handle by default.)

> Instead I think we should write it like `IterableNode`

The annoying thing about using `instanceof` is that the compiler can't use this information for anything useful. Sure, `IterableNode` looks nice and simple, but that's because it does very little. Consider if there were eight different kinds of nodes that we wanted to count as `IterableNode`. In `getAnElement`, we would then have an eight-way disjunction, with a branch for each kind of "internal" node, casting and dispatching to the appropriate predicate. This is very anti-modular. (And if you happen to forget a branch of the disjunct, then it just silently fails. Oof.)

Anyway, in this case it's irrelevant, since `target` might not be a `SequenceNode` at all. In that case, I would prefer an explicit `instanceof` check at this point, rather than an inline cast (even if this ultimately does the same thing).